pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+                                                            block_arg)) {
+            reqs_met = false;
+            break;
+          }
+        } catch (const std::out_of_range &e) {
+          IVLOG(1, "Error message: " << e.what());
+          IVLOG(1, "Requested key was: " << std::make_pair(i, curr_idx));
+          throw;
+        }
+      }
+      if (!reqs_met) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      llvm::SmallVector<mlir::BlockArgument, 8> new_bound_idxs(bound_idxs);
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
+  {
+    auto maybe_loads_and_stores = capture();
+    if (maybe_loads_and_stores) {
+      loads_and_stores = maybe_loads_and_stores.getValue();
+    } else {
+      IVLOG(4, "Cannot Stencil: Operations fail to pattern-match.");
+      return;
+    }
+  }
+
+  llvm::SmallVector<mlir::Value, 3> tensors;
+  for (auto &load_op : loads_and_stores.loads) {
+    // TODO: It's probably fine, but I'm a little worried `getMemRef` isn't
+    // const
+    tensors.push_back(load_op.getMemRef());
+  }
+  size_t first_store_idx = tensors.size();
+  for (auto &store_op : loads_and_stores.stores) {
+    // TODO: It's probably fine, but I'm a little worried `out` isn't const
+    tensors.push_back(store_op.out());
+    // TODO: Probably should handle reduces vs. true stores in a different way
+    // if (auto reduce_op = llvm::dyn_cast_or_null<AffineReduceOp>(store_op)) {
+    //   tensors.push_back(reduce_op.out());
+    // } else if (auto true_store_op =
+    // llvm::dyn_cast_or_null<mlir::AffineStoreOp>(store_op)) {
+    //   tensors.push_back(true_store_op.getMemRef());
+    // } else {
+    //   // TODO: throw?
+    //   IVLOG(1, "Unexpected failure to load tensors from ops in stenciling");
+    //   return;
+    // }
+  }
+  std::sort(tensors.begin(), tensors.begin() + first_store_idx);
+  do { // Each load tensor permutation
+    std::sort(tensors.begin() + first_store_idx, tensors.end());
+    do { // Each store tensor permutation
+      // Add all legal permutations to legal_permutations
+      RecursiveBindIndex(llvm::SmallVector<mlir::BlockArgument, 8>(), tensors);
+    } while (std::next_permutation(tensors.begin() + first_store_idx,
+                                   tensors.end()));
+  } while (std::next_permutation(tensors.begin(),
+                                 tensors.begin() + first_store_idx));
+
+  // TODO: If we desire more complex requirements than pairwise tensor-to-index
+  // stride requirements, that function could go here
+
+  for (const auto &perm : legal_permutations) {
+    // TODO: Clean, try to get everything on the same unsigned type/width
+    auto ranges = op.getConstantRanges();
```

Comment:
This is constant over the entire search (i.e. it could even be done in the constructor, or certainly at the dawn of DoStenciling if you are worried about failure during constructions.  This call is not trivial in terms of the work required.

Reply:
:+1: I had a TODO that I must have accidentally deleted that I was doing that in the wrong place, thanks for flagging.